### DISCUSSION TITLE
Queries Related to Quantum Number Wrapper, Basis State Construction for Custom Hamiltonian

### DISCUSSION THREAD

**Discussion by JessicaJohnBritto** [block-hczhai/block2-preview#117] [Q&A] [answered]
**Hamiltonian**: 1D Bose Hubbard Model. Fixed number of particles, no spin degrees of freedom.
Basis States are in the occupation number basis. For eg, in case N = 2 sites, k = 2 bosonic particles in the entire system, the basis states are (2, 0), (1, 1), (0, 2) --- where (x1, x2) - means x1 no. of particles on site-1 and x2 no. of particles on site-2, and their sum should be equal to `k`, i.e, `sum(occupation number basis) = k` --- `k-constraint`.
`n_max` - max no of particles each site can hold. 

I have been trying to follow the example given in [Custom Hamiltonian section](https://block2.readthedocs.io/en/latest/tutorial/custom-hamiltonians.html#The-Hubbard-Model) to use it for my model.

1. Could you please tell me what this `1` after the Q-tuple means in this `(Q(0, 0, 0), 1)`? 
2. Can I generate all the possible states in a list and pass it to Q?
3. What does `SX` do in `Q = driver.bw.SX`? I understand bw - is the object of Block2Wrapper.
4. In the `driver.initialize_system`  - how can I impose k-constraint on `Q(n_max, n_max)`?
5. In this code - `b.add_term("cd", np.array([i for i in range(L) for _ in range(4)]), U)`, is 4 being used since for each site for the hopping term, when it is flattened, we get 4 elements, so to keep the length of the vector same for Hopping and onsite interaction terms - `site_index` is being repeated 4 times for each site?

I am getting `AssertionError` in each of the following lines 

- `basis = [(Q(2, 0), 1), (Q(1, 1), 1), (Q(2, 0), 1)] # [0ab2]`
- `driver.initialize_system(n_sites=L, vacuum=Q(0, 0), target=Q(n_max, n_max), hamil_init=False)`

Here's the [link to the notebook](https://github.com/JessicaJohnBritto/MTP/blob/main/Initial%20Codes/block2_DMRG-Copy1.ipynb) for this case.

Thanks

**Comment by hczhai** [comment, marked as answer]
I have added an example script for the Bose-Hubbard model in the updated [documentation](https://block2.readthedocs.io/en/latest/tutorial/custom-hamiltonians.html#Bose-Hubbard-Model).

> Could you please tell me what this 1 after the Q-tuple means in this (Q(0, 0, 0), 1)?

The ``1`` is number of states per quantum number, which is explained in the [documentation](https://block2.readthedocs.io/en/latest/tutorial/custom-hamiltonians.html#The-Hubbard-Model): "The quantum number and number of states in each quantum number at each site (which defines the local Hilbert space) is set in ``site_basis``."

> I am getting AssertionError in each of the following lines

To solve these problems, please read the documentation carefully. For example, the [documentation](https://block2.readthedocs.io/en/latest/tutorial/custom-hamiltonians.html#Custom-Symmetry-Groups) says: "The number of arguments in the quantum number wrapper ``Q`` should then match the number of sub-group names given in ``DMRGDriver.set_symmetry_groups``."

> In this code - ``b.add_term("cd", np.array([i for i in range(L) for _ in range(4)]), U)``, is 4 being used since for each site for the hopping term

This line of code is not part of the documentation.

**Reply by JessicaJohnBritto** [reply]
Yes, it's cleared now. Thank you!
